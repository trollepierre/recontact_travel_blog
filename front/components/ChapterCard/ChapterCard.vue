<template>
  <div class="chapter-card">
    <article class="chapter">
      <header class="chapter__header">
        <h2 class="chapter__title">
          {{ chapterTitle }}
        </h2>
        <p class="chapter__position">
          {{ chapter.position }}
        </p>
      </header>
      <div class="chapter__content">
        <img
          v-if="imgLink"
          :src="imgLink"
          :alt="chapterAlt"
          class="chapter__image">
        <span v-else>
          {{ $t("missingImage") }}
        </span>
      </div>
      <footer class="chapter__footer">
        <div
          v-for="paragraph in chapterText"
          :key="paragraph.text"
          class="chapter__footer_text">
          <template v-if="paragraph">
            <a
              v-if="paragraph.isLink"
              :href="paragraph.text"
              target="_blank">
              {{ paragraph.text }}
            </a>
            <p v-else>
              {{ paragraph.text }}
            </p>
          </template>
        </div>
      </footer>
    </article>
  </div>
</template>

<script>
  import translationsService from '../../services/services/translations'

  export default {
    name: 'ChapterCard',
    props: {
      chapter: { type: Object, default: () => {} },
    },
    computed: {
      imgLink() {
        const { imgLink } = this.chapter
        return !imgLink ? false : imgLink
      },
      chapterTitle() {
        const language = this.$store.state.locale
        return translationsService.getChapterTitle(this.chapter, language)
      },
      chapterAlt() {
        return this.$t('altComplement') + this.chapterText[0].text
      },
      chapterText() {
        const language = this.$store.state.locale
        const chapterText = translationsService.getChapterText(this.chapter, language)

        return chapterText
          .filter(paragraph => !!paragraph)
          .map(paragraph => {
            let isLink = false
            /* eslint-disable no-useless-escape */
            const urlRegExp = new RegExp('^(https?:\/\/)([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?')
            if (urlRegExp.test(paragraph)) {
              isLink = true
            }
            return { isLink, text: paragraph }
          })
      },
    },
    i18n: {
      messages: {
        fr: {
          missingImage: 'Image manquante',
          altComplement: 'Une image montrant ',
        },
        en: {
          missingImage: 'Missing image',
          altComplement: 'An image showing ',
        },
      },
    },
  }
</script>

<style scoped>
  h2 {
    font-weight: normal;
  }

  img.chapter__image {
    max-width: 100%;
    max-height: 40vw;
  }

  @media only screen and (max-width: 640px) {
    img.chapter__image {
      max-height: 100%;
    }
  }

  .chapter {
    min-width: 260px;
    /*max-width: 260px;*/
    background: #ffffff;
    border-radius: 4px !important;
    box-shadow: 0 1px 1px rgba(0, 0, 0, .15);
    border: 1px solid rgba(0, 0, 0, .09);
    display: flex;
    flex-direction: column;
    color: #535a60;
  }

  .chapter__header {
    border-bottom: 1px solid #e6e6e6;
    padding: 15px;
    height: 34px;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .chapter__title {
    font-size: 16px;
    font-weight: 700;
    line-height: 17px;
    color: #07c;
    margin: 0;
    overflow-wrap: break-word;
  }

  .chapter__position {
    display: none;
  }

  .chapter__content {
    font-size: 15px;
    padding: 15px;
    display: block;
    color: #000;
    text-align: center;
  }

  .chapter__footer {
    text-align: center;
    padding: 15px;
    border-top: 1px solid #e6e6e6;
  }

  .chapter__footer button {
    text-transform: uppercase;
    color: #d14800;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 10px;
    font-weight: 700;
  }

  .chapter__footer button:hover {
    background: #d14800;
    color: #ffffff;
  }

  .chapter__footer button:disabled,
  .chapter__footer button:active {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

</style>
