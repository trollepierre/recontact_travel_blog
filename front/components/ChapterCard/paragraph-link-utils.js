const urlTester = text => /^(https?:\/\/)?([\da-z.-]+)\.([a-z.]{2,6})([/\w .-]*)*\/?/.test(text)

const youtubeEmbedUrlTester = text => urlTester(text) && (/youtu.be/.test(text) || /youtube.com/.test(text)) && /embed/.test(text)
const generateIframeLink = paragraph => `${paragraph}?rel=0&modestbranding=1`
const generateCleanUrlLink = paragraph => paragraph.includes('https://') ? paragraph : `https://${paragraph}`

export {
  youtubeEmbedUrlTester,
  generateCleanUrlLink,
  generateIframeLink,
  urlTester,
}
