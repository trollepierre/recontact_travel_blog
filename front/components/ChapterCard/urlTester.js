export const urlTester = text => {
  const urlRegExp = new RegExp('^(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?') /* eslint-disable-line no-useless-escape */
  return urlRegExp.test(text)
}

export const youtubeEmbedUrlTester = text => urlTester(text) && (/youtu.be/.test(text) || /youtube.com/.test(text)) && /embed/.test(text)
