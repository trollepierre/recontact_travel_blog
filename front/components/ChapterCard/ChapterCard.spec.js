import VueI18n from 'vue-i18n'
import Vuex from 'vuex'
import ChapterCard from './ChapterCard.vue'
import translationsService from '../../services/services/translations'

describe('Component | ChapterCard.vue', () => {
  let wrapper
  const chapter = {
    position: 82,
    frTitle: 'Titre du premier paragraphe',
    enTitle: 'Title of the first paragraph',
    imgLink: 'webf',
    frText: 'du blabla',
    enText: 'some blabla',
  }
  let localVue
  let store

  beforeEach(() => {
    console.warn = jest.fn()
    translationsService.getChapterTitle = jest.fn()
    translationsService.getChapterTitle.mockReturnValue('My title')
    translationsService.getChapterText = jest.fn()
    translationsService.getChapterText.mockReturnValue(['Title of the first paragraph', 'https://github.com/trollepierre', 'url', 'https://www.youtube.com/embed/-18AYp_7iW0'])
    localVue = createLocalVue()
    localVue.use(VueI18n)
    localVue.use(Vuex)
    const propsData = { chapter }
    store = new Vuex.Store({ state: { locale: 'en' } })
    wrapper = shallowMount(ChapterCard, { localVue, propsData, store })
  })

  it('should be named "ChapterCard"', () => {
    expect(wrapper.name()).toEqual('ChapterCard')
  })

  describe('template', () => {
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should match snapshot when imgLink is not set', () => {
      chapter.imgLink = ''
      const propsData = {
        chapter,
      }
      wrapper = shallowMount(ChapterCard, { localVue, propsData, store })
      expect(wrapper).toMatchSnapshot()
    })

    it('should not contain header when chapterTitle is empty', () => {
      translationsService.getChapterTitle.mockReturnValue('-')
      const propsData = {
        chapter,
      }
      wrapper = shallowMount(ChapterCard, { localVue, propsData, store })
      expect(wrapper.find('.chapter__header').element).toBeUndefined()
    })
  })

  describe('computed property #imgLink', () => {
    it('should return imgLink when defined', () => {
      chapter.imgLink = 'dropbox.com/img0.jpg'
      const propsData = { chapter }

      wrapper = shallowMount(ChapterCard, { localVue, propsData, store })

      expect(wrapper.vm.imgLink).toEqual('dropbox.com/img0.jpg')
    })

    it('should return false when chapter imglink is empty', () => {
      chapter.imgLink = ''
      const propsData = { chapter }

      wrapper = shallowMount(ChapterCard, { localVue, propsData, store })

      expect(wrapper.vm.imgLink).toEqual(false)
    })
  })

  describe('locales', () => {
    const languages = Object.keys(ChapterCard.i18n.messages)

    it('contains 2 languages', () => {
      expect(languages).toHaveLength(2)
      expect(languages).toEqual(['fr', 'en'])
    })

    describe('each language', () => {
      describe('fr', () => {
        const locales = Object.keys(ChapterCard.i18n.messages.fr)

        it('contains 2 locales', () => {
          expect(locales).toHaveLength(2)
          expect(locales).toEqual([
            'missingImage',
            'altComplement',
          ])
        })
      })

      describe('en', () => {
        const locales = Object.keys(ChapterCard.i18n.messages.en)

        it('contains 2 locales', () => {
          expect(locales).toHaveLength(2)
          expect(locales).toEqual([
            'missingImage',
            'altComplement',
          ])
        })
      })
    })
  })
})

