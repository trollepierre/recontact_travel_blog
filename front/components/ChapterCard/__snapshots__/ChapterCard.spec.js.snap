// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | ChapterCard.vue template should match snapshot 1`] = `
<div
  class="chapter-card"
>
  <article
    class="chapter"
  >
    <header
      class="chapter__header"
    >
      <h2
        class="chapter__title"
      >
        
        My title
      
      </h2>
       
      <p
        class="chapter__position"
      >
        
        82
      
      </p>
    </header>
     
    <div
      class="chapter__content"
    >
      <img
        alt="altComplementone text"
        class="chapter__image"
        src="webf"
      />
    </div>
     
    <footer
      class="chapter__footer"
    >
      <div
        class="chapter__footer_text"
      >
        <p>
          
            one text
          
        </p>
      </div>
    </footer>
  </article>
</div>
`;

exports[`Component | ChapterCard.vue template should match snapshot when imgLink is not set 1`] = `
<div
  class="chapter-card"
>
  <article
    class="chapter"
  >
    <header
      class="chapter__header"
    >
      <h2
        class="chapter__title"
      >
        
        My title
      
      </h2>
       
      <p
        class="chapter__position"
      >
        
        82
      
      </p>
    </header>
     
    <div
      class="chapter__content"
    >
      <span>
        
        missingImage
      
      </span>
    </div>
     
    <footer
      class="chapter__footer"
    >
      <div
        class="chapter__footer_text"
      >
        <p>
          
            one text
          
        </p>
      </div>
    </footer>
  </article>
</div>
`;
